src/platform/forms-system/src/js/fields/FileField.jsx

```diff
@@ -29,25 +39,123 @@ export default class FileField extends React.Component {
+      // keep - file may contain File object; ensure name isn't empty
+      // remove - file may be an empty object
+      data => !data.file || (data.file?.name || '') !== '',
+    );
+    if (formData.length !== (this.props.formData || []).length) {
+      this.props.onChange(formData);
+    }
+  }
+
+  focusAddAnotherButton = () => {
+    // focus on span pseudo-button, not the label
+    focusElement(`#${this.props.idSchema.$id}_add_label span`);
+  };
+
+  onSubmitPassword = (file, index, password) => {
+    if (file && password) {
+      this.onAddFile({ target: { files: [file] } }, index, password);
+    }
+  };
+
+  isFileEncrypted = async (file, index) => {
+    const { onChange, formData } = this.props;
+    return checkForEncryptedPdf(file)
+      .then(isEncrypted => {
+        if (isEncrypted) {
+          onChange(
+            _.set(
+              index,
+              { file, name: file.name, index, isEncrypted: true },
+              formData || [],
+            ),
+          );
+        }
+        return isEncrypted;
+      })
+      .catch(() => {
+        // This _should_ only be called if a file is deleted after the
+        // user selects it for upload
+        this.setFileState({ index, isEncrypted: false });
+        onChange(
+          _.set(
+            index,
+            { file, name: file.name, index, isEncrypted: false },
+            formData || [],
+          ),
+        );
+        return false;
+      });
+  };
+
+  /**
+   * Add file to list and upload
+   * @param {Event} event - DOM File upload event
+   * @param {number} index - uploaded file index, if already uploaded
+   * @param {string} password - encrypted PDF password, only defined by
+   *   `onSubmitPassword` function
+   * @listens
+   */
+  onAddFile = async (event, index = null, password) => {
     if (event.target.files && event.target.files.length) {
+      const currentFile = event.target.files[0];
       const files = this.props.formData || [];
+      const { requestLockedPdfPassword, uiSchema } = this.props;
+      const { getEncryptedPassword } = uiSchema['ui:options'] || {};
+
       let idx = index;
       if (idx === null) {
         idx = files.length === 0 ? 0 : files.length;
       }
+
+      // Check if the file is an encrypted PDF
+      if (
+        requestLockedPdfPassword && // feature flag
+        getEncryptedPassword && // uiSchema flag
+        currentFile.name?.endsWith('pdf') &&
+        !password
+      ) {
+        const needsPassword = await this.isFileEncrypted(currentFile, idx);
+        if (needsPassword) {
+          // wait for user to enter a password before uploading
+          this.props.onChange(
+            _.set(
+              idx,
+              {
+                file: currentFile,
+                name: currentFile.name,
+                isEncrypted: true,
+              },
+              this.props.formData || [],
+            ),
+          );
+          return;
+        }
```

Comment:
Why are we calling `onChange` twice? The `isFileEncrypted` call will call it if the file is encrypted, and once we return true we call it again here.

Reply:
Good catch! Added a single call to `onChange` above the `if (needsPassword) {` check..